### PR TITLE
GH-1407: Support Symbol.hasInstance on interfaces (proposal)

### PR DESCRIPTION
**Pull request by dbo**
#1407 

@mor-n4 please

### CODE REVIEW

**Comment by mor-n4** [comment]
Nice idea!

I'll try to look into the corresponding transpiler changes within the next few days.

File: n4js-libs/packages/n4js-runtime/src/js/rt/N4RuntimeBootstrap.js [hunk 1]
```diff
@@ -13,30 +13,18 @@
 (function (global) {
     "use strict";
 
-    var ArraySlice = Array.prototype.slice,
+    var symHasInstance = Symbol.hasInstance,
+        ArraySlice = Array.prototype.slice,
         noop = function() {};
 
-    if (typeof __REACT_HOT_LOADER__ !== "undefined") {
```
Comment (mor-n4) [inline n4js-libs/packages/n4js-runtime/src/js/rt/N4RuntimeBootstrap.js]: What was this for? Is this obsolete?

File: n4js-libs/packages/n4js-runtime/src/js/rt/N4RuntimeBootstrap.js [hunk 2]
```diff
@@ -257,8 +244,12 @@
     global.$makeClass = $makeClass;
     global.$makeInterface = $makeInterface;
     global.$makeEnum = $makeEnum;
-    global.$implements = $implements;
-    global.$instanceof = $instanceof;
+
+    // @deprecated, remove:
+    global.$implements = $implements;    
+    // @deprecated, remove:
+    global.$instanceof = function(instance, supertype) { return instance instanceof supertype; };
```
Comment (mor-n4) [inline n4js-libs/packages/n4js-runtime/src/js/rt/N4RuntimeBootstrap.js]: if we merge this soon, I can remove those as part of `GH-1485`

File: n4js-libs/packages/n4js-runtime/src/js/rt/N4RuntimeBootstrap.js [hunk 3]
```diff
@@ -64,7 +52,7 @@
      *                                interfaces defined in definition files without annotation @N4JS.
      * @param instanceMethods - An object holding the methods for the class instance and mixed in methods
      * @param staticMethods - An object holding the descriptors for the class static methods
-     * @param n4typeFn - The factory function to create the meta type.
+     * @param n4typeFn - Optional factory function to create the meta type.
```
Comment (mor-n4) [inline n4js-libs/packages/n4js-runtime/src/js/rt/N4RuntimeBootstrap.js]: As we discussed yesterday, we cannot make reflection optional before we had a look at the classes (i.e. trying to use ES6 classes). Making this parameter optional right now will make it more difficult to understand what's going on when looking at the classes without having a true benefit right now. Or am I overlooking something?

File: n4js-libs/packages/n4js-runtime/src/js/rt/N4RuntimeBootstrap.js (hunk 1 above)
Comment (dbo) [reply n4js-libs/packages/n4js-runtime/src/js/rt/N4RuntimeBootstrap.js]: is outdated, could go

File: n4js-libs/packages/n4js-runtime/src/js/rt/N4RuntimeBootstrap.js (hunk 3 above)
Comment (dbo) [reply n4js-libs/packages/n4js-runtime/src/js/rt/N4RuntimeBootstrap.js]: No, this was just a prework to be able to omit reflection in the future.

**Comment by mor-n4** [comment]
THanks @dbo, merging ...

**Pull request merged**